staging/kos/pkg/controllers/subnet/validator.go

```diff
@@ -0,0 +1,458 @@
+	subnetRef := k8stypes.NamespacedName{
+		Namespace: s.Namespace,
+		Name:      s.Name,
+	}
+	v.queue.Add(subnetRef)
+}
+
+func (v *Validator) OnSubnetUpdate(oldObj, newObj interface{}) {
+	oldS, newS := oldObj.(*netv1a1.Subnet), newObj.(*netv1a1.Subnet)
+	glog.V(5).Infof("Notified of update from %#+v to %#+v.", oldS, newS)
+	subnetRef := k8stypes.NamespacedName{
+		Namespace: newS.Namespace,
+		Name:      newS.Name,
+	}
+
+	// Process a subnet only if the fields that affect validation have changed.
+	if oldS.Spec.IPv4 != newS.Spec.IPv4 || oldS.Spec.VNI != newS.Spec.VNI {
+		v.queue.Add(subnetRef)
+	}
+}
+
+func (v *Validator) OnSubnetDelete(obj interface{}) {
+	s := parse.Peel(obj).(*netv1a1.Subnet)
+	glog.V(5).Infof("Notified of deletion of %#+v.", s)
+	subnetRef := k8stypes.NamespacedName{
+		Namespace: s.Namespace,
+		Name:      s.Name,
+	}
+	v.queue.Add(subnetRef)
+}
+
+func (v *Validator) processQueue() {
+	for {
+		subnet, stop := v.queue.Get()
+		if stop {
+			return
+		}
+		v.processQueueItem(subnet.(k8stypes.NamespacedName))
+	}
+}
+
+func (v *Validator) processQueueItem(subnet k8stypes.NamespacedName) {
+	defer v.queue.Done(subnet)
+	requeues := v.queue.NumRequeues(subnet)
+	if err := v.processSubnet(subnet); err != nil {
+		glog.Warningf("Failed processing %s, requeuing (%d earlier requeues): %s.", subnet, requeues, err.Error())
+		v.queue.AddRateLimited(subnet)
+		return
+	}
+	glog.V(4).Infof("Finished %s with %d requeues.", subnet, requeues)
+	v.queue.Forget(subnet)
+}
+
+func (v *Validator) processSubnet(subnetNSN k8stypes.NamespacedName) error {
+	subnet, err := v.subnetLister.Subnets(subnetNSN.Namespace).Get(subnetNSN.Name)
+
+	if err != nil && !k8serrors.IsNotFound(err) {
+		glog.Errorf("subnet lister failed to lookup %s: %s", subnetNSN, err.Error())
+		// This should never happen. No point in retrying.
+		return nil
+	}
+
+	if k8serrors.IsNotFound(err) {
+		v.processDeletedSubnet(subnetNSN)
+		return nil
+	}
+
+	return v.processExistingSubnet(subnet)
+}
+
+func (v *Validator) processDeletedSubnet(s k8stypes.NamespacedName) {
+	v.clearStaleRV(s)
+
+	rivals := v.clearConflictsCache(s)
+
+	// Enqueue old rivals so that they can be re-validated: they might no longer
+	// have conflicts as this subnet has been deleted.
+	for _, r := range rivals {
+		v.queue.Add(r)
+	}
+}
+
+func (v *Validator) processExistingSubnet(s *netv1a1.Subnet) error {
+	ss, parsingErrs := subnet.NewSummary(s)
+	if len(parsingErrs) > 0 {
+		return parsingErrs
+	}
+
+	if v.subnetIsStale(ss.NamespacedName, s.ResourceVersion) {
+		return nil
+	}
+
+	// If we're here s might have been created or updated in a way that affects
+	// validation. We need to update its conflicts cache accordingly and
+	// reconsider old rivals because they might no longer be in conflict with
+	// s.
+	oldRivals := v.updateConflictsCache(ss)
+	for _, r := range oldRivals {
+		v.queue.Add(r)
+	}
+
+	// When a subnet is validated, in case of conflicts it is added to the
```

Comment:
What is the relevance of the first two sentences here?

I think this can and should be greatly shrunk.  Yes, in SubnetStatus document that the apiserver clears `Validated` when `Spec.VNI` or `Spec.IPv4` changes.  The comment on this early out can be something simple like
```
// Keep the promise that a Subnet stays validated once it becomes validated (unless and until
// its VNI or IP range changes).

Reply:
Fixed in 8ac9883.